README.md
Электронное зажигание мотоцикла ИЖ Планета 5.
Управление опережением зажигания в зависимости от оборотов двигателя, вывод текущей частоты вращения двигателя на дисплей с интерфейсом ILI9341.

Микроконтроллер MicroChip PIC16F1619. Выполнено в среде MPLab.

Конфигурации:
SOFT - отладка в среде MPLab без использования SPI интерфейса для вывода информации на LED экран.
TEST - отладка на микроконтроллере с программной эмуляцией работы датчика Холла.
HARD - для прошивки реального утсройства.

Electronic Ignition For IZ Planeta 5


History.txt
05.01.2020

Удален таймер TMR0, заменен на TMR1.
На таймере TMR1 реализуется прерывание 60 мкс.
Выствлен период прерывания 59.82 мкс, что соответствует обработке
прерывания с периодом 60 мкс.


05.04.2020

Выход RB4 является выходом для SPI, но так же является входом для FUNC2. Закомментировал
снятие сигнала с RB4 в main.c.
Убран сигнал SCREEN_RESET. Вход на дисплее подтянут к питанию +5V.

04.04.2020

Переносим сюда таблицу смещения искры

/**
 Auxiary function
 Table jumps
 */

/**
ORG	0x0300

TableDefault;

    addwf   PCL,F               ; Вычисляемый переход

    retlw   .000                ; 0
    retlw   .000                ; 1
    retlw   .000                ; 2
    retlw   .000                ; 3
    retlw   .000                ; 4
    retlw   .000                ; 5
    retlw   .000                ; 6
    retlw   .000                ; 7
    retlw   .000                ; 8
    retlw   .000                ; 9
    retlw   .000                ; 10
    retlw   .000                ; 11
    retlw   .000                ; 12
    retlw   .000                ; 13
    retlw   .000                ; 14
    retlw   .000                ; 15
    retlw   .007                ; 16
    retlw   .008                ; 17
    retlw   .009                ; 18
    retlw   .009                ; 19
    retlw   .010                ; 20
    retlw   .010                ; 21
    retlw   .011                ; 22
    retlw   .012                ; 23
    retlw   .012                ; 24
    retlw   .013                ; 25
    retlw   .014                ; 26
    retlw   .015                ; 27
    retlw   .015                ; 28
    retlw   .016                ; 29
    retlw   .017                ; 30
    retlw   .018                ; 31
    retlw   .019                ; 32
    retlw   .019                ; 33
    retlw   .020                ; 34
    retlw   .021                ; 35
    retlw   .022                ; 36
    retlw   .022                ; 37
    retlw   .023                ; 38
    retlw   .024                ; 39
    retlw   .025                ; 40
    retlw   .025                ; 41
    retlw   .026                ; 42
    retlw   .027                ; 43
    retlw   .028                ; 44
    retlw   .029                ; 45
    retlw   .030                ; 46
    retlw   .031                ; 47
    retlw   .031                ; 48
    retlw   .032                ; 49
    retlw   .033                ; 50
    retlw   .034                ; 51
    retlw   .035                ; 52
    retlw   .036                ; 53
    retlw   .037                ; 54
    retlw   .037                ; 55
    retlw   .038                ; 56
    retlw   .039                ; 57
    retlw   .040                ; 58
    retlw   .041                ; 59
    retlw   .042                ; 60
    retlw   .043                ; 61
    retlw   .044                ; 62
    retlw   .045                ; 63
    retlw   .046                ; 64
    retlw   .046                ; 65
    retlw   .047                ; 66
    retlw   .048                ; 67
    retlw   .049                ; 68
    retlw   .050                ; 69
    retlw   .051                ; 70
    retlw   .052                ; 71
    retlw   .053                ; 72
    retlw   .054                ; 73
    retlw   .054                ; 74
    retlw   .055                ; 75
    retlw   .056                ; 76
    retlw   .057                ; 77
    retlw   .058                ; 78
    retlw   .059                ; 79
    retlw   .060                ; 80
    retlw   .061                ; 81
    retlw   .062                ; 82
    retlw   .062                ; 83
    retlw   .063                ; 84
    retlw   .064                ; 85
    retlw   .065                ; 86
    retlw   .066                ; 87
    retlw   .067                ; 88
    retlw   .068                ; 89
    retlw   .069                ; 90
    retlw   .070                ; 91
    retlw   .071                ; 92
    retlw   .072                ; 93
    retlw   .073                ; 94
    retlw   .074                ; 95
    retlw   .075                ; 96
    retlw   .076                ; 97
    retlw   .077                ; 98
    retlw   .078                ; 99
    retlw   .079                ; 100
    retlw   .080                ; 101
    retlw   .081                ; 102
    retlw   .082                ; 103
    retlw   .082                ; 104
    retlw   .083                ; 105
    retlw   .084                ; 106
    retlw   .085                ; 107
    retlw   .086                ; 108
    retlw   .087                ; 109
    retlw   .088                ; 110
    retlw   .089                ; 111
    retlw   .090                ; 112
    retlw   .091                ; 113
    retlw   .092                ; 114
    retlw   .093                ; 115
    retlw   .094                ; 116
    retlw   .095                ; 117
    retlw   .096                ; 118
    retlw   .097                ; 119
    retlw   .098                ; 120
    retlw   .099                ; 121
    retlw   .100                ; 122
    retlw   .101                ; 123
    retlw   .102                ; 124
    retlw   .102                ; 125
    retlw   .103                ; 126
    retlw   .104                ; 127
    retlw   .105                ; 128
    retlw   .106                ; 129
    retlw   .107                ; 130
    retlw   .108                ; 131
    retlw   .109                ; 132
    retlw   .110                ; 133
    retlw   .111                ; 134
    retlw   .112                ; 135
    retlw   .113                ; 136
    retlw   .114                ; 137
    retlw   .115                ; 138
    retlw   .116                ; 139
    retlw   .117                ; 140
    retlw   .118                ; 141
    retlw   .119                ; 142
    retlw   .120                ; 143
    retlw   .121                ; 144

main.h
/* Microchip Technology Inc. and its subsidiaries.  You may use this software 
 * and any derivatives exclusively with Microchip products. 
 * 
 * THIS SOFTWARE IS SUPPLIED BY MICROCHIP "AS IS".  NO WARRANTIES, WHETHER 
 * EXPRESS, IMPLIED OR STATUTORY, APPLY TO THIS SOFTWARE, INCLUDING ANY IMPLIED 
 * WARRANTIES OF NON-INFRINGEMENT, MERCHANTABILITY, AND FITNESS FOR A 
 * PARTICULAR PURPOSE, OR ITS INTERACTION WITH MICROCHIP PRODUCTS, COMBINATION 
 * WITH ANY OTHER PRODUCTS, OR USE IN ANY APPLICATION. 
 *
 * IN NO EVENT WILL MICROCHIP BE LIABLE FOR ANY INDIRECT, SPECIAL, PUNITIVE, 
 * INCIDENTAL OR CONSEQUENTIAL LOSS, DAMAGE, COST OR EXPENSE OF ANY KIND 
 * WHATSOEVER RELATED TO THE SOFTWARE, HOWEVER CAUSED, EVEN IF MICROCHIP HAS 
 * BEEN ADVISED OF THE POSSIBILITY OR THE DAMAGES ARE FORESEEABLE.  TO THE 
 * FULLEST EXTENT ALLOWED BY LAW, MICROCHIP'S TOTAL LIABILITY ON ALL CLAIMS 
 * IN ANY WAY RELATED TO THIS SOFTWARE WILL NOT EXCEED THE AMOUNT OF FEES, IF 
 * ANY, THAT YOU HAVE PAID DIRECTLY TO MICROCHIP FOR THIS SOFTWARE.
 *
 * MICROCHIP PROVIDES THIS SOFTWARE CONDITIONALLY UPON YOUR ACCEPTANCE OF THESE 
 * TERMS. 
 */

/* 
 * File:   
 * Author: 
 * Comments:
 * Revision history: 
 */

// This is a guard condition so that contents of this file are not included
// more than once.  
#ifndef XC_HEADER_TEMPLATE_H
#define	XC_HEADER_TEMPLATE_H

#include <xc.h> // include processor files - each processor file is guarded.
#include <stdio.h>
#include "mcc_generated_files/test_engine_rpm.h"

// TODO Insert appropriate #include <>

// TODO Insert C++ class definitions if appropriate

// TODO Insert declarations
// Store state of toggle
typedef union {
    unsigned char all;
    struct {
        unsigned SENS                       :1; //Hall sense state
        unsigned NOTDEF1                    :1; //Nodef1
        unsigned NOTDEF2                    :1; //Nodef2
        unsigned NOTDEF3                    :1; //Nodef3
        unsigned SELECT1                    :1; //Select table ignition #1
        unsigned SELECT2                    :1; //Select table ignition #2
        unsigned FUNC1                      :1; //Engine stop
        unsigned FUNC2                      :1; //Warm spark
    };
} STATE_TOGGLE_t;

STATE_TOGGLE_t Port = {0x00};

// Store flags
typedef union {
    unsigned char all;
    struct {
        unsigned lastState                    :1; //Last state Hall sensor
        unsigned overflowCount                    :1; //Overflow count
        unsigned coilOff                      :1; //Coil disable when engine not work
        unsigned engineStop                      :1; //Disable engine
        unsigned overrun1                    :1; //Enable limit overrun FUNC1
        unsigned overrun2                    :1; //Enable limit overrun FUNC2
        unsigned NOTDEF1                    :1; //Nodef1
        unsigned NOTDEF2                    :1; //Nodef2
    };
} FLAG_t;

FLAG_t Flag = {0x00};



uint8_t countSELECT1 = 0;
uint8_t countSELECT2 = 0;
uint8_t countFUNC1 = 0;
uint8_t countFUNC2 = 0;
uint8_t countHALL = 0; //счетчик фильтра датчика Холла
uint8_t sectorCount = 0; //Счетчик отсчетов прерываний по таймеру при проходе шторки через датчик Холла
uint8_t lastSectionCount = 0; //Последнее значение счетчик отсчетов прерываний по таймер при проходе шторки через датчик Холла
uint8_t sparkTime = 0; //Кол-во отсчетов после которого происходит включение катушки при нахождении шторки в датчике Холла
const uint8_t countEnought = 255;
const uint8_t countHallEnought = 4;
uint16_t coilCount = 0; //Счетчик включения катушки
uint16_t coilOffCount = 0; //Счетчик простоя катушки

// Table shift ignition
uint8_t shiftIgnMassive[] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                            7, 8, 9 ,9 ,10 ,10, 11, 12, 12, 13, 14, 15, 15, 16, 17, 18,
                            19, 19, 20, 21, 22, 22, 23, 24, 25, 25, 26, 27, 28, 29, 30, 31,
                            31, 32, 33, 34, 35, 36, 37, 37, 38, 39, 40, 41, 42, 43, 44, 45,
                            46, 46, 47, 48, 49, 50, 51, 52, 53, 54, 54, 55, 56, 57, 58, 59,
                            60, 61, 62, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74,
                            75, 76, 77, 78, 79, 80, 81, 82, 82, 83, 84, 85, 86, 87, 88, 89,
                            90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 102, 103, 104,
                            105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118,
                            119, 120, 121, 122, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131,
                            132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 141, 143, 144, 146,
                            147, 148, 150, 151, 152, 154, 155, 156, 158, 159, 160, 162, 163, 164,
                            166, 167, 168, 170, 171, 172, 174, 175, 177, 178, 179, 181, 182, 183,
                            185, 186, 187, 189, 190, 191, 193, 194, 195, 197, 198, 199, 201, 202,
                            203, 205, 206, 208, 209, 210, 212, 213, 214, 216, 217, 218, 220, 221,
                            222, 224, 225, 226, 228, 229, 230, 232, 233, 234, 236, 237, 238, 000};

const uint8_t secPerMin = 60; // Seconds per minutes

// Comment a function and leverage automatic documentation with slash star star
/**
    <p><b>Function prototype:</b></p>
  
    <p><b>Summary:</b></p>

    <p><b>Description:</b></p>

    <p><b>Precondition:</b></p>

    <p><b>Parameters:</b></p>

    <p><b>Returns:</b></p>

    <p><b>Example:</b></p>
    <code>
 
    </code>

    <p><b>Remarks:</b></p>
 */
// TODO Insert declarations or function prototypes (right here) to leverage 
// live documentation

#ifdef	__cplusplus
extern "C" {
#endif /* __cplusplus */

    // TODO If C++ is being used, regular C code needs function names to have C 
    // linkage so the functions can be used by the c code. 

#ifdef	__cplusplus
}
#endif /* __cplusplus */

#endif	/* XC_HEADER_TEMPLATE_H */

